[LoopIdiom] Select llvm.experimental.memset.pattern intrinsic rather than memset_pattern16 libcall

asb opened the pull request:
*Note*: This patch is fully ready for technical review, but is not ready for merging until ongoing testing that the altered codegen doesn't (as far as can be seen) regress code generation for the existing libcall, which may result in other patches that should land before this one. I post this for feedback now, because any alternate implementation approach would impact this testing.

---

In order to keep the change as incremental as possible, this only introduces the memset.pattern intrinsic in cases where memset_pattern16 would have been used. Future patches can enable it on targets that don't have the intrinsic, and select it in cases where the libcall isn't directly usable. As the memset.pattern intrinsic takes the number of times to store the pattern as an argument unlike memset_pattern16 which takes the number of bytes to write, we no longer try to form an i128 pattern.

Special care is taken for cases where multiple stores in the same loop iteration were combined to form a single pattern. For such cases, we inherit the limitation that loops such as the following are supported:

```
for (unsigned i = 0; i < 2 * n; i += 2) {
  f[i] = 2;
  f[i+1] = 2;
}
```

But the following doesn't result in a memset.pattern (even though it could be, by forming an appropriate pattern):
```
for (unsigned i = 0; i < 2 * n; i += 2) {
  f[i] = 2;
  f[i+1] = 3;
}
```

Addressing this existing deficiency is left for a follow-up due to a desire not to change too much at once (i.e. to target equivalence to the current codegen).

A command line option is introduced to force the selection of the intrinsic even in cases it wouldn't be (i.e. in cases where the libcall wouldn't have been selected). This is intended as a transitionary option for testing and experimentation, to be removed at a later point.

asb committed:
[LoopIdiom] Select llvm.experimental.memset.pattern intrinsic rather than memset_pattern16 libcall

In order to keep the change as incremental as possible, this only
introduces the memset.pattern intrinsic in cases where memset_pattern16
would have been used. Future patches can enable it on targets that don't
have the intrinsic. As the memset.pattern intrinsic takes the number of
times to store the pattern as an argument unlike memset_pattern16 which
takes the number of bytes to write, we no longer try to form an i128
pattern.

Special care is taken for cases where multiple stores in the
same loop iteration were combined to form a single pattern. For such
cases, we inherit the limitation that loops such as the following are
supported:

```
for (unsigned i = 0; i < 2 * n; i += 2) {
  f[i] = 2;
  f[i+1] = 2;
}
```

But the following doesn't result in a memset.pattern (even though it
could be, by forming an appropriate pattern):
```
for (unsigned i = 0; i < 2 * n; i += 2) {
  f[i] = 2;
  f[i+1] = 3;
}
```

Addressing this existing deficiency is left for a follow-up due to a
desire not to change too much at once (i.e. to target equivalence to the
current codegen).

A command line option is introduced to force the selection of the
intrinsic even in cases it wouldn't be (i.e. in cases where the libcall
wouldn't have been selected). This is intended as a transitionary option
for testing and experimentation, to be removed at a later point.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Fix bit/bytes confusion in setting alignment

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Excise unneeded logic for creating a globalvariable (which was still done for global pointers)

We can just do a ptrtoint.

arsenm commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1154,7 +1154,7 @@ bool LoopIdiomRecognize::processLoopStridedStore(
       GV->setUnnamedAddr(
           GlobalValue::UnnamedAddr::Global); // Ok to merge these.
-      GV->setAlignment(Align(PatternArgTy->getPrimitiveSizeInBits()));
+      GV->setAlignment(Align(PatternArgTy->getPrimitiveSizeInBits() / 8));
```
Just above you already queried the getTypeSizeInBits for PatternArgTy, avoid repeating it? Every getPrimitiveSizeInBits makes me nervous

asb replied:
The logic to create a GlobalVariable isn't needed any more, so I've now dropped this whole bit.

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1064,53 +1073,95 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+  // MemsetArg is the number of bytes for the memset libcall, and the number
+  // of pattern repetitions if the memset.pattern intrinsic is being used.
+  Value *MemsetArg;
+  std::optional<int64_t> BytesWritten = std::nullopt;
```
Do we need `= std::nullopt`?

asb replied:
No, I prefer being explicit but I guess this is more a personal preference and not consistently done within LLVM.

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1064,53 +1073,95 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+            : Builder.CreateMul(TripCount,
+                                Builder.getIntN(IntIdxTy->getIntegerBitWidth(),
+                                                PatternRepsPerTrip));
+    if (auto CI = dyn_cast<ConstantInt>(TripCount))
```
I believe this should be `auto *CI` by coding standards.

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1064,53 +1073,95 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+      return Changed;
+    MemsetArg =
+        Expander.expandCodeFor(NumBytesS, IntIdxTy, Preheader->getTerminator());
+    if (auto CI = dyn_cast<ConstantInt>(MemsetArg))
```
`auto *CI`

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1064,53 +1073,95 @@ bool LoopIdiomRecognize::processLoopStridedStore(
-    assert (isLibFuncEmittable(M, TLI, LibFunc_memset_pattern16));
+    assert(ForceMemsetPatternIntrinsic ||
+           isLibFuncEmittable(M, TLI, LibFunc_memset_pattern16));
     // Everything is emitted in default address space
```
This comment about address space no longer applies

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Remove unnecessary initialisation of std::optional

asb committed:
auto *CI and remove comment that's made redundant

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Set alignment for dest ptr of emitted memset.pattern

asb committed:
Tweak description of ForceMemsetPatternIntrinsic

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -303,10 +308,15 @@ bool LoopIdiomRecognize::runOnLoop(Loop *L) {
+  // TODO: Unconditionally enable use of the memset pattern intrinsic (or at
+  // least, opt-in via target hook) once we are confident it will never result
+  // in worse codegen than without. For now, use it only when we would have
+  // previously emitted a libcall to memset_pattern16 (or unless this is
```
Instead of referring to "previously", should we say "For now, use it only when the target supports memset_pattern16 libcall (or unless this is overridden by command line option)."

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Drop outdated comment

asb committed:
Tweak comment as per Craig's suggestion

asb commented:
> Reverse ping. Can you rebase this?

Just catching up post EuroLLVM. I've resolved the merge conflicts now, and fixed and outdated comment and accepted your suggestion (thanks!). And now putting some more focused effort in getting this closed off.

I have a flow for looking for unexpected codegen differences (which should only show up on darwin where the libcall is supported) by applying a hacky patch that unconditionally marks memset_pattern16 as available on all platforms, then building the test-suite with `-k0` (to ignore link errors) and looking at assembly diffs. Going through this, there are a few things I'm looking at that need resolution before this can be merged:
* More minor
  * Duplicated globals vs before. The linker can clean it up, and tracing through it's just because of things like inlining, where before we made one global in LoopIdiomRecognize and later may have cloned the call that uses it. But now, we may clone the intrinsic call and then create a global for each one that is lowered to memset_pattern16.
  * Loosing some type information in the emitted globals (bit patterns are the same, but before commented as 'float' etc appropriately). I've not looked at this as it's not a functional change.
  * In some examples we no longer tailcall memset_pattern16. Marking as minor as I didn't see a case it's likely to make a big difference, but it would be nice to clean up
* The bigger one is I see some examples with repeated memset_pattern16 calls that wasn't there before. After looking at some of the others, I'm now isolating a smaller example.

It's possible some of the above results in changes to this patch, but also equally possible they result in separate PRs that end up being effectively pre-requisites for this one.

asb committed:
Fix incorrectly resolved merge conflict

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

preames commented:
> The bigger one is I see some examples with repeated memset_pattern16 calls that wasn't there before. After looking at some of the others, I'm now isolating a smaller example.

I wrote a bunch of test cases to see if I could identity your problem.  I believe this is a consequence of the second parameter (the value param) not being known to be readnone.  There might be other cases, but I can definitely see this case failing to fold in DSE and GVN.

preames commented:
First attempt at a partial fix for the pointer value operand issue above: https://github.com/llvm/llvm-project/pull/138559

preames commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1083,53 +1092,92 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+  std::optional<int64_t> BytesWritten;
+
+  if (PatternValue && (HasMemsetPattern || ForceMemsetPatternIntrinsic)) {
+    const SCEV *TripCountS =
```
There's a bunch of complexity in this section of the change that I'm not getting.  Since we need the NumBytes (in the original code) for the AAInfo, why not just leave that alone?  Isn't count (for memset.pattern intrinsic)  just the trip count of the loop by definition?

preames replied:
I still think this might be reworked, but we're down to little enough code here I'm okay with that being a followup.

asb replied:
Yes, it's the case of multiple stores in a loop iteration. I could perhaps build the necessary value with numbytes and a div, but I'm wary of cases where it may not be folded away.

preames commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1083,53 +1092,92 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+        SE->getTripCountFromExitCount(BECount, IntIdxTy, CurLoop);
+    if (!Expander.isSafeToExpand(TripCountS))
+      return Changed;
+    const SCEVConstant *ConstStoreSize = dyn_cast<SCEVConstant>(StoreSizeSCEV);
```
Maybe it helps to point out that StoreSizeSCEV is just the scev of a TypeSize and we could pass that in here instead?

preames commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1083,53 +1092,92 @@ bool LoopIdiomRecognize::processLoopStridedStore(
+    // If ConstStoreSize is not equal to the width of PatternValue, then
+    // MemsetArg is TripCount * (ConstStoreSize/PatternValueWidth). Else
+    // MemSetArg is just TripCount.
+    MemsetArg =
```
Isn't this just weird way of compute NumBytes?

preames commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -1419,7 +1467,7 @@ bool LoopIdiomRecognize::processLoopStoreOfLoopLoad(
   AAMDNodes StoreAATags = TheStore->getAAMetadata();
   AATags = AATags.merge(StoreAATags);
-  if (auto CI = dyn_cast<ConstantInt>(NumBytes))
+  if (auto *CI = dyn_cast<ConstantInt>(NumBytes))
```
precommit this please

preames reviewed (commented):
FYI, I reorganized the code a bit, so you'll have to rebase.  Feel free to defer until after answering my high level question though - I'm suspecting we can simplify this a bit.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb commented:
Just a holding pattern update to explain the last push:
I've merged in the latest changes and adjusted to match the upstream. I've pushed so we can work from a relatively up to date base.
* I know my changes slightly undo some of the refactoring Philip did. I will revisit.
* I do still see some unwanted/unexpected changes and am looking to understand those at the moment.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Make use of the fact memset.pattern allows any sized type as arg

asb committed:
[PreISelIntrinsicLowering] Reuse previously generated GlobalVariable for memset_pattern16 when possible

As Constants are already uniquified, we can use a map to keep track of
whether a GlobalVariable was produced for a given Constant or not.
Repeated globals with the same value was one of the codegen differences
noted in #126736. This patch removes that diff, producing cleaner
output.

asb commented:
I've further reduced diffs and investigated the changes that were unexpected and am satisfied there's nothing untoward going on:
* (Minor) We were always casting to integer types. Now that the memset.pattern intrinsic was expanded to allow any sized type this is unnecessary. Deleted the code that did this casting, and diffs vs the old codepath are reduced as generated globals in the output .s now have comments reflecting the type.
* We were previously making a new GlobalVariable for every memset_pattern16 in a Module. I've added a map that tracks constants that have already been emitted and avoid this. Again, this reduces diff when comparing to the current codegen path. I've cherry-picked the patch here, but it's posted as an independent PR https://github.com/llvm/llvm-project/pull/144677
* The remaining unexpected codegen diff I was seeing was again around unrolling. This time, TTI `isLoweredToCall` being called from `getUnrollingPreferences`. This returns `false` with the new codegen path but not the old one, affecting unrolling. For local testing I have a minimal patch that teaches `isLoweredToCall` about the memset.pattern intrinsic but don't intend to upstream it (the method has very little coverage of intrinsics as-is and there doesn't seem to be a justification in adding just memset.pattern).

TODO: re-review initial patch in light of some review comments and upstream refactorings.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

preames commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -150,6 +150,11 @@ static cl::opt<bool> UseLIRCodeSizeHeurs(
              "with -Os/-Oz"),
     cl::init(true), cl::Hidden);
 
+static cl::opt<bool> ForceMemsetPatternIntrinsic(
```
I would mildly suggest dropping this cl::opt from the final patch.  We probably should have a generic means to force enable a TLI function, but having one only for memset.pattern seems slightly odd.  Not a strongly held view, will leave to your discretion.

arsenm replied:
I do have a patch somewhere that adds an opt flag to do this

arsenm replied:
https://github.com/llvm/llvm-project/pull/145808

preames commented on llvm/test/Transforms/LoopIdiom/basic.ll:
```diff
@@ -596,7 +594,7 @@ for.end:                                          ; preds = %for.body
 ; CHECK-LABEL: @test13_pattern(
 ; CHECK-NEXT:  entry:
-; CHECK-NEXT:    call void @memset_pattern16(ptr [[P:%.*]], ptr @.memset_pattern.1, i64 80000)
+; CHECK-NEXT:    call void @llvm.experimental.memset.pattern.p0.p0.i64(ptr align 4 [[P:%.*]], ptr @G, i64 10000, i1 false)
```
Just to note, you dropped the pointer to int conversion from the prior version of the patch, and thus until a reworked https://github.com/llvm/llvm-project/pull/138559 lands we have an aliasing imprecision here.  I'm explicitly okay with that being true for the landed patch, and will follow up on my patch post-commit.

preames reviewed (approved):
LGTM

nikic commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -411,14 +421,12 @@ static Constant *getMemSetPatternValue(Value *V, const DataLayout *DL) {
-  if (Size == 16)
-    return C;
+  // For now, don't handle types that aren't int, floats, or pointers.
+  if (!isa<ConstantInt>(C) && !isa<ConstantFP>(C) &&
```
Better check the type for these, otherwise it's going to implicitly allow vector splats in the future...

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q1-loopidiom-recognize-switch-to-memset-pattern-intrinsic

asb committed:
Check type directly in getMemsetPatternValue as suggested by nikic

asb committed:
Fix comment at top of getMemSetPatternValue

asb commented:
> LLVM Buildbot has detected a new failure on builder `clang-hip-vega20` running on `hip-vega20-0` while building `llvm` at step 3 "annotate".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/123/builds/22990

A subsequent build https://lab.llvm.org/buildbot/#/builders/123/builds/22992 passed, so I think there is not a failure introduced by this patch.
